sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/FileIndex.scala

```diff
@@ -72,4 +72,14 @@ trait FileIndex {
 
   /** Schema of the partitioning columns, or the empty schema if the table is not partitioned. */
   def partitionSchema: StructType
+
+  /**
+   * Returns an optional metadata operation time, in nanoseconds, for listing files.
+   *
+   * We do file listing in query optimization (in order to get the proper statistics) and we want
+   * to account for file listing time in physical execution (as metrics). To do that, we save the
+   * file listing time in some implementations and physical execution calls it in this method
+   * to update the metrics.
+   */
+  def metadataOpsTimeNs: Option[Long] = None
```

Comment:
I think it's hard to define the semantic of this method for general `FileIndex`, as everytime we call `listFiles`, the value of this method should be updated.

how about we only put this method in `PrunedInMemoryFileIndex`?